iotfunctions/anomaly.py

```diff
@@ -1227,50 +1212,442 @@ def __init__(self, input_item, windowsize, output_item):
+                datatype=float,
+                description="Data item to analyze"
+            )
+        )
+
+        inputs.append(
+            UISingle(
+                name="windowsize",
+                datatype=int,
+                description="Size of each sliding window in data points. Typically set to 12.",
+            )
+        )
+
+        # define arguments that behave as function outputs
+        outputs = []
+        outputs.append(
+            UIFunctionOutSingle(
+                name="output_item",
+                datatype=float,
+                description="Anomaly score (SaliencybasedGeneralizedAnomalyScore)",
+            )
+        )
+        return (inputs, outputs)
+
+
+#######################################################################################
+# Anomaly detectors with scaling
+#######################################################################################
+class KMeansAnomalyScorev2(Standard_Scaler):
+    '''
+    An unsupervised anomaly detection function.
+     Applies a k-means analysis clustering technique to time series data.
+     Moves a sliding window across the data signal and applies the anomaly model to each window.
+     The window size is typically set to 12 data points.
+     The normalize switch allows to learn and apply a standard scaler prior to computing the anomaly score.
+     Try several anomaly models on your data and use the one that fits your databest.
+    '''
+    eval_metric = staticmethod(metrics.r2_score)
+
+    # class variables
+    train_if_no_model = True
+
+    def __init__(self, input_item, windowsize, normalize, output_item, expr=None):
+        super().__init__(features=[input_item], targets=[output_item], predictions=None)
+
+        logger.debug(input_item)
+        # do not run score and call transform instead of predict
+
+        self.input_item = input_item
+
+        # use 12 by default
+        self.windowsize, windowoverlap = set_window_size_and_overlap(windowsize)
+
+        # step
+        self.step = self.windowsize - windowoverlap
+
+        self.normalize = normalize
+
+        # assume 1 per sec for now
+        self.frame_rate = 1
+
+        self.output_item = output_item
+
+        self.whoami = 'KMeansv2'
+
+
+    def prepare_data(self, dfEntity):
+
+        logger.debug(self.whoami + ': prepare Data')
+
+        # operate on simple timestamp index
+        if len(dfEntity.index.names) > 1:
+            index_names = dfEntity.index.names
+            dfe = dfEntity.reset_index().set_index(index_names[0])
+        else:
+            index_names = None
+            dfe = dfEntity
+
+        # interpolate gaps - data imputation
+        try:
+            dfe = dfe.interpolate(method="time")
+        except Exception as e:
+            logger.error('Prepare data error: ' + str(e))
+
+        # one dimensional time series - named temperature for catchyness
+        temperature = dfe[[self.input_item]].fillna(0).to_numpy().reshape(-1,)
```

Comment:
Currently we are not applying the anomaly function on the scaled data, with the same random_state both this version and the old `KMeansAnomalyScore` return the same scores . `self.input_item` should be the replaced with the scaled/transformed input, which in our case is `'predicted_'+self.output.item` (but coded better)
```python3
temperature = dfe[['predicted_'+self.output.item]].fillna(0).to_numpy().reshape(-1,)
```

Reply:
good catch. Thank you very much for spotting this
I forgot to make sure I'm using the "predicted" (scaled) data for scoring after reorganizing the code.